pep-0683.rst

```diff
@@ -0,0 +1,335 @@
+than comparing with ``_Py_IMMORTAL_REFCNT`` directly.
+
+(There are other minor, internal changes which are not described here.)
+
+This is not meant to be a public feature but rather an internal one.
+So the proposal does *not* including adding any new public C-API,
+nor any Python API.  However, this does not prevent us from
+adding (publicly accessible) private API to do things
+like immortalize an object or tell if one
+is immortal.
+
+Affected API
+------------
+
+API that will now ignore immortal objects:
+
+* (public) ``Py_INCREF()``
+* (public) ``Py_DECREF()``
+* (public) ``Py_SET_REFCNT()``
+* (private) ``_Py_NewReference()``
+
+API that exposes refcounts (unchanged but may now return large values):
+
+* (public) ``Py_REFCNT()``
+* (public) ``sys.getrefcount()``
+* (public) ``sys.gettotalrefcount()``
+
+Immortal Global Objects
+-----------------------
+
+The following objects will be made immortal:
+
+* singletons (``None``, ``True``, ``False``, ``Ellipsis``, ``NotImplemented``)
+* all static types (e.g. ``PyLong_Type``, ``PyExc_Exception``)
+* all static objects in ``_PyRuntimeState.global_objects`` (e.g. identifiers,
+  small ints)
+
+There will likely be others we have not enumerated here.
+
+Documentation
+-------------
+
+The feature itself is internal and will not be added to the documentation.
+
+We *may* add a note about immortal objects to the following,
+to help reduce any surprise users may have with the change:
+
+* ``Py_SET_REFCNT()`` (a no-op for immortal objects)
+* ``Py_REFCNT()`` (value may be surprisingly large)
+* ``sys.getrefcount()`` (value may be surprisingly large)
+
+Other API that might benefit from such notes are currently undocumented.
+
+We wouldn't add a note anywhere else (including for ``Py_INCREF()`` and
+``Py_DECREF()``) since the feature is otherwise transparent to users.
+
+
+Rejected Ideas
+==============
+
+Equate Immortal with Immutable
+------------------------------
+
+Making a mutable object immortal isn't particularly helpful.
+The exception is if you can ensure the object isn't actually
+modified again.  Since we aren't enforcing any immutability
+for immortal objects it didn't make sense to emphasis
+that relationship.
+
+
+Reference Implementation
+========================
+
+The implementation is proposed on GitHub:
+
+https://github.com/python/cpython/pull/19474
+
+
+Open Issues
+===========
+
+* how do we ensure all immortal objects get cleaned up during runtime finalization?
+* how do we adjust ``sys.gettotalrefcount()`` to reflect things properly (for the sake of buildbots)?
+* is there any other impact on GC?
+
+
+References
+==========
+
+.. [python-dev] This was discussed in December 2021 on python-dev:
+   https://mail.python.org/archives/list/python-dev@python.org/thread/7O3FUA52QGTVDC6MDAV5WXKNFEDRK5D6/#TBTHSOI2XRWRO6WQOLUW3X7S5DUXFAOV
+   https://mail.python.org/archives/list/python-dev@python.org/thread/PNLBJBNIQDMG2YYGPBCTGOKOAVXRBJWY
+
```

Comment:
@ericsnowcurrently I noticed on my last read through that this footnote isn't referenced anywhere anymore. If intentional, I'd suggest making it body text in the references section and not a footnote.

A